### PR TITLE
Introduce handler for .docx files

### PR DESCRIPTION
**Pull request by igavriil**
Problem and/or solution
-----------------------

With this commit a new handler for handling .docx files
is introduced. A high level overview is described below.

The docx parsing is based on ISO/IEC 29500-1:2016 [https://www.iso.org/standard/71691.html]
that describes the office open xml file formats. In a nutshell
a docx file is a zipped directory that contains a set of xml files.

The folder/file structured expected when unzipping a .docx file is:
```
        /_rels/
            .rels
            [Content_Types].xml
            ...
        /word/
            document.xml
            /_rels/
                document.xml.rels
            ...
```

The two xmls that will use to extract translatable content are:
* /word/document.xml -> Contains all(*) the textual content of the document
* /word/_rels/document.xml.rels -> Contains relationships details linked
  from /word/document.xml. These relationships will be used to extract
  external hyperlinks that are associated with text elements in the document

The string segmentation follows the segmentation of the textual content
into pararagraphs as this is defined inside document.xml.

A simplified version of the two files:
* word/document.xml
```
<w:document>
    <w:body>
        <w:p>
            ...
            <w:r>
                ...
                <w:t xml:space="preserve">hello</w:t>
                ...
            </w:r>
            ...
            <w:r>
                ...
                <w:t xml:space="preserve">world</w:t>
                ...
            </w:r>
            ...
        </w:p>
        ...
	<w:p>
            <w:hyperlink r:id="rId6">
                ...
                <w:r>
                     ...
                     <w:t xml:space="preserve">this is a link</w:t>
                     ...
                </w:r>
                ...
            </w:hyperlink>
        </w:p>
    </w:body>
</w:document>
```

* word/_rels/document.xml.rels
```
<Relationships>
    ...
    <Relationship Id="rId6" Target="https://www.transifex.com/" TargetMode="External"/>
    ...
</Relationships>
```

Parsing
-------
* All paragraphs (w:p elements) are inspected
* For each paragraph all child text elements (w:t) are combined
  into a single source string (seperated by pseudo tags <tx>)
  - For text elements surrounded by a hyperlink (w:hyperlink) the
    pseudo tag also includes the href extracted from `document.xml.rels`
* In each paragraph we add a special attribute (`txid`) which contains
  the hash of the final extracted source string. This will be helpful
  when compiling back the .docx file.

Complilation
------------
* All paragraphs (w:p elements) are inspected
* For each paragraph extract the attribute (`txid`) and find the
  corresponding translation by hash from the stringset provided
* Split the translation into parts by separating it at pseudo tags (<tx>)
* For each child text element (w:t) in the paragraph replace the text
  with the appropriate part of the translation

Tricks:
* There are some text elements that contain only spaces. In order to
  reduce noise from the extracted source string, we ignore them
  and prepend additional whitespace to the next text element.
  The reverse process is also taking place on complilation where
  we detect these text elements and we remove trailing spaces
  from the parts that need to.

* When the number of the text elements in a paragraph are not equal
  to the translation parts of the translation, we either:
  * remove text elements from the xml
  * combine multiple translation parts into one

How to test
-----------
Use as a guide the unit tests in order
to check how parsing and compilation behaves.

Reviewer checklist
------------------

Code:
* [ ] Change is covered by unit-tests
* [ ] Code is well documented, well styled and is following [best practices](https://tem.transifex.com)
* [ ] Performance issues have been taken under consideration
* [ ] Errors and other edge-cases are handled properly

PR:
* [ ] Problem and/or solution are well-explained
* [ ] Commits have been squashed so that each one has a clear purpose
* [ ] Commits have a proper commit message [according to TEM](https://tem.transifex.com/github-guide.html#working-on-a-feature)

### CODE REVIEW

File: openformats/formats/docx.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        os.mkdir(self.__tmp_folder)
+
+        docx_path = '{}/{}.docx'.format(self.__tmp_folder, 'in')
+        with io.open(docx_path, 'wb') as f:
+            f.write(content)
+
+        with ZipFile(docx_path, 'r') as z:
+            z.extractall(self.__tmp_folder)
+
+        self.__filelist = z.namelist()
```
Comment (kbairak) [inline openformats/formats/docx.py:97]: What's with the privacy? :slightly_smiling_face:

File: openformats/formats/docx.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        as `txid` attribute in order to be able to match when
+        compilation takes place.
+        """
+        docx = DocxFile(content)
+
+        soup = BeautifulSoup(docx.get_document(), 'xml')
+        rels_soup = BeautifulSoup(docx.get_document_rels(), 'xml')
+
+        stringset = []
+        order = 1
```
Comment (kbairak) [inline openformats/formats/docx.py:215]: Minor, you can use `itertools.count` for this. It would also be more consistent with the rest of the handlers.

**Review by kbairak — COMMENTED** [review]
The formatting ends up a bit messed up. If parts of a pargraph are styled, they end up having the style of the first part of the paragaph. Also, if a picture is anchored as a character, it gets misplaced in the translation:

Screenshot from 2021-02-11 12-41-37
Screenshot from 2021-02-11 12-41-28

I guess these issues can't be helped though...

I'm going to fix some flake8 issues and then approve this.

**Review by kbairak — APPROVED** [review]
